### DISCUSSION TITLE
web_accessible_resources resources are not compiled if they are typescript files

### DISCUSSION THREAD

**Discussion by woss** [crxjs/chrome-extension-tools#616] [Q&A]
Following code snippet is the part of the manifest setting

```js
  "content_scripts": [
    {
      "js": [
        "src/content.ts"
      ],
      "matches": [
        "http://*/*",
        "https://*/*"
      ],
      "run_at": "document_start"
    }
  ],
 "web_accessible_resources": [
    {
      "resources": [
        "src/myPageToInject.ts"
      ],
      "matches": [
        "http://*/*",
        "https://*/*"
      ],
      "use_dynamic_url": true
    }
  ],
```
Yields something like this: 
```js
  "web_accessible_resources": [
    {
      "matches": [
        "http://*/*",
        "https://*/*"
      ],
      "resources": [
        "src/myPageToInject.ts",
        "assets/content.ts-b571cc48.js"
      ],
      "use_dynamic_url": true
    }
  ],
```

The in the `dist` is still unprocessed TS file. Is this a bug or a missing feature?

**Comment by insilications** [comment]
It would be nice to have this feature.

**Reply by Toumash** [reply]
https://github.com/crxjs/chrome-extension-tools/discussions/616#discussioncomment-16528061

**Comment by woss** [comment]
i agree, i was going through the code and indeed there is no compilation process for the `web-accessible-resources`

@jacksteamdev how hard would this be to implement?

**Reply by Toumash** [reply]
https://github.com/crxjs/chrome-extension-tools/discussions/616#discussioncomment-16528061

**Comment by YunFeng0817** [comment]
@jacksteamdev I also really want this feature. Is there any plan to support this in the future?

**Reply by Toumash** [reply]
https://github.com/crxjs/chrome-extension-tools/discussions/616#discussioncomment-16528061

**Comment by flexchar** [comment]
This repo is probably dead... :(

**Reply by woss** [reply]
i wonder is he alive? also i wonder why creators of popular repos don't write something about the lack of their time so others can contribute or at least know not to use the project since it might not receive updates or new features.

**Reply by flexchar** [reply]
I'm a developer myself. Sometimes the life aligns in such a way that the next weekend where I thought I'm gonna sit down and contribute never arrives. Open Source also doesn't pay the bills as easily (although I'm more than open supporting the painful work it does and abstracts from us). I do agree on the lack of communication. It feels like the current README preaches as this being the ultimate winner but the commit history hits below the waist. 

Last night I found this. Looks very promising but a whole new thing to learn. I wish the core team working on the Manifest 3 had considered developers and HMR feature for local development. If not reloading, many of us would probably stick to existing tools. :)

https://www.plasmo.com
https://github.com/PlasmoHQ/plasmo

**Reply by woss** [reply]
unfortunately, the OSS doesn't pay the bills for 98% of the OSS creators. I started a project some time ago that might have changed that but it didn't took off, yet. 

do you mind telling me why did you consider moving from plasma(if you are using it) to this? 

i could use some alternatives :)

**Reply by flexchar** [reply]
I actually moved FROM ParcelJS to this to get proper HMR working on content scripts. But now I need a script in the `MAIN` world and this doesn't support. So I am considering moving TO Plasma as it seems to support everything one can imagine (looking at their examples [repo here](https://github.com/PlasmoHQ/examples)). 

They're also a commercial product and 30usd/mon is more than fine if it saves me a couple of hours and lowers mental complexity! But I have not tried yet.

**Reply by woss** [reply]
thanks for the info, i will keep this in mind :)

**Comment by beautyfree** [comment]
Same needs

**Comment by flexchar** [comment]
I keep getting notifications... I wanted to say I moved to Plasmo framework. It works wonders. It doesn't have proper HMR for Vue but it allows to do so much more already. I would definitely advise to give a shot if the code base is not big yet.

**Comment by MindsightsAI** [comment]
ANy updates on this? How to solve that problem? Maybe anybody know workarounds?

**Reply by jan-wilhelm** [reply]
From my experience I haven’t found any work around and simply resorted to writing web accessible resources in plain JavaScript

**Comment by b5414** [comment]
Instead of _web_ass_, you can do a trick with _the module background_ + `scripting.executeScript`

**Reply by revel111** [reply]
Hello, could you elaborate more on that?

**Comment by KajKandler** [comment]
This seems still to be an issue in April 2026.

**Comment by Toumash** [comment]
first time seeing this discussion, sorry.
It looks like an attempt to use the library in other way it was designed. Basically one of the foundations in crxjs is that the WAR's are generated, not taken from the user. 
I'm not sure what do the OP is trying to achieve here, an example would be a good start

**Reply by Toumash** [reply]
After digging into the code, I think the OP's use case is injecting a TypeScript file into the page's MAIN world from a content script -  the classic pattern of creating a <script> tag with chrome.runtime.getURL() and appending it to the DOM.
They listed the .ts file directly in web_accessible_resources because that's what Chrome's own docs tell you to do. But in CRXJS, WAR entries are treated as static assets to copy, not scripts to compile - so the raw .ts file ends up in dist/ unprocessed.
The good news is this is already supported via the `?script` import query. Instead of listing the file in web_accessible_resources manually, you do:
```ts
// src/content.ts
import scriptUrl from './myPageToInject?script&module'
const script = document.createElement('script')
script.src = chrome.runtime.getURL(scriptUrl)
document.head.appendChild(script)
```

CRXJS will compile the TypeScript, add the output to web_accessible_resources automatically, and give you the correct URL.
There are existing tests for this pattern in packages/vite-plugin/tests/out/vite-declared-script-resources/ and packages/vite-plugin/tests/out/dynamic-script/.
So this is more of a discoverability gap than a missing feature - users coming from vanilla Chrome extension development expect .ts files in web_accessible_resources to be compiled the same way content_scripts entries are.
